### DISCUSSION TITLE
Set pump speed/rpm directly from MQTT?

### DISCUSSION THREAD

**Discussion by MaxVonEvil** [tagyoureit/nodejs-poolController#526] [Q&A]
@tagyoureit @bsileo and everyone else
Hi, is it possible to set the active RPM value for a pump directly from MQTT? 
(No Pentair controller here, just a Pi4 running NjsPC 7.7.0 and a IntelliFlo VS pump here,). 

I got as far as defining a set of 3 different circuits:
Nnamed them 1000, 2000 and 3450 for the RPM's they represent
Assigned those circuits to the pump in the dashpanel, setting the actual RPM value to match. 

I can now hit the command topic nixie-singlebody/state/circuits/setstate  with a payload of {"id":X,"isOn":true}, X being the circuit. However as to my question; is there a way to do something like 'nixie-singlebody/state/pumps/setstate' with payload {"id":50,"rpm":NNNN}, thus not having to define circuits assigned predefined speeds?

Thanks a bunch for a great project

**Comment by tagyoureit** [comment]
No, this isn't possible.  The pump circuits are assigned pump speeds and the pump takes the max speed of all circuits that are currently on.  However, the pump will continually evaluate these in real time.  If you send an API request to change all of the pump circuits (yes, they all need to be sent together) the pump will pick up the new speed even if it is already running.

eg, sending the following api request:
```
{
    "method": "put",
    "url": "/config/pump",
    "data": "{\"id\":50,\"circuits\":[{\"speed\":3400,\"units\":{\"val\":0},\"id\":1,\"circuit\":8},{\"speed\":2450,\"units\":{\"val\":0},\"id\":2,\"circuit\":7}]}"
}
```

Will set pump circuit 1, assigned to id 8, to speed 3400 and pump circuit 2, assigned to circuit 7 to pump speed 2450.

**Reply by MaxVonEvil** [reply]
Thank you sir! Okay I guess I can make this work. In my scenario I was able to shave the above down to one circuit, then overwriting the speed in that circuit as I need. 
I don't suppose this works via MQTT as well?

As I wanted to avoid talking both REST and MQTT, I decided to move all comms to RESt. Found out I needed the ability to start/stop the circuit too. Found the [API guide](https://tagyoureit.github.io/nodejs-poolcontroller-api/#/state/put_state_circuit_setState_), and added this:

method.put
url: http://njspc-host:4200/state/circuit/setState
payload: {"id":6,"isOn":**true or false**}
(using node-red, i.e. no need to escape the doublequotes etc)

**Reply by tagyoureit** [reply]
Why do you want to control the pump in this manner?  And how many speeds do you think you would want to control the pump?

**Reply by MaxVonEvil** [reply]
Well, I am experimenting with valve actuators, waterfalls and laminar flow nozzles (a guy's gotta have a hobby, right? ;)

Being able to precisely dial-in the RPM's in an easy way, is super convenient. Based on the info you provided above I was able to create a slider in Home-Assistant where I can seamlessly set the pump speed.

I guess the only thing I need MQTT now for is subscribing to the watts and rpm values coming back from the pump as constantly polling for updates via REST seems inefficient. 

PS: I don't suppose there's a way to reconfigure the RPM of the pump circuit via MQTT? My OCD is hissing at me for using both REST and MQTT, lol

**Reply by rstrouse** [reply]
> I guess the only thing I need MQTT now for is subscribing to the watts and rpm values coming back from the pump as constantly polling for updates via REST seems inefficient.

Or you could subscribe to a socket and it will hand it to you.  Not sure if there is a socket client in HASS though.

EDIT:  Or you could interface the njsPC socket to call a rest command on HASS as well.

**Comment by MaxVonEvil** [comment]
@rstrouse - Thanks for the suggestion, I do not have a whole lot of experience subscribing to sockets, although I can see there's a built in WebSockets node in Node Red, which I'm using as the brains of this setup. 

Is it a wss://my-njspc-host:port-something/somethingelse, or does one configure a listener and point njsPC to it? If you can point me in the general direction of what a sockets configuration looks like for njsPC it would be highly appreciated, thanks!

**Reply by tagyoureit** [reply]
The websockets are on the same port as njsPC - 4200.  If you load up dashPanel and open the developer console you can see all of the events that are being returned.  

![image](https://user-images.githubusercontent.com/1646840/179335995-2be2cebc-23b2-4389-b2f2-78b772ab2c9d.png)

The data structures are the same as you'll find in the API returns for similar event names.  (They are documented [here](https://tagyoureit.github.io/nodejs-poolcontroller-api/), but 2 years old so need a refresh.)

**Reply by MaxVonEvil** [reply]
@tagyoureit, that is very useful, yet I do feel a bit silly here, as I was unaware of said dev console in the dashpanel? How/where does one access it?

**Reply by tagyoureit** [reply]
The developer console is in Chrome.  [Open the developer tools](https://developer.chrome.com/docs/devtools/open/) and then click click the console.
![image](https://user-images.githubusercontent.com/1646840/179363450-3b567415-dd5c-4720-9671-efecb0eb292b.png)

**Reply by MaxVonEvil** [reply]
Ohgod, I am facepalming myself so hard right now 😆 
Anyway thanks a bunch - I'll see what I can make of this!
